src/main/java/com/google/api/codegen/transformer/StandardImportSectionTransformer.java

```diff
@@ -24,25 +25,48 @@
 
 public class StandardImportSectionTransformer implements ImportSectionTransformer {
   @Override
-  public ImportSectionView generateImportSection(TransformationContext context) {
-    return generateImportSection(context.getImportTypeTable().getImports());
+  public ImportSectionView generateImportSection(TransformationContext context, String className) {
+    return generateImportSection(context.getImportTypeTable().getImports(), className);
   }
 
   @Override
   public ImportSectionView generateImportSection(
       MethodContext context, Iterable<InitCodeNode> specItemNodes) {
-    return generateImportSection(context.getTypeTable().getImports());
+    return generateImportSection(context.getTypeTable().getImports(), null);
   }
 
-  public ImportSectionView generateImportSection(Map<String, TypeAlias> typeImports) {
+  public ImportSectionView generateImportSection(
+      Map<String, TypeAlias> typeImports, String className) {
     ImmutableList.Builder<ImportFileView> appImports = ImmutableList.builder();
     for (TypeAlias alias : typeImports.values()) {
+      if (excludeAppImport(alias, className)) {
+        continue;
+      }
       ImportTypeView.Builder imp = ImportTypeView.newBuilder();
       imp.fullName(alias.getFullName());
       imp.nickname(alias.getNickname());
       imp.type(alias.getImportType());
+
       appImports.add(ImportFileView.newBuilder().types(ImmutableList.of(imp.build())).build());
     }
     return ImportSectionView.newBuilder().appImports(appImports.build()).build();
   }
+
+  private boolean excludeAppImport(TypeAlias alias, String className) {
+    String parentName = alias.getParentFullName();
+    if (className == null
+        || alias.getImportType() != ImportType.StaticImport
+        || parentName == null
+        || !parentName.endsWith(className)) {
+      return false;
+    }
+
+    if (parentName.length() > className.length()) {
+      char packageSeparator = parentName.charAt(parentName.length() - className.length() - 1);
+      if (Character.isLetterOrDigit(packageSeparator)) {
+        return false;
+      }
+    }
```

Comment:
This is java-specific logic and thus should probably be in a java-specific class.

Reply:
Added comment as we discussed offline. Also included underscore as a valid classname character.